File: packages/analytics/service.py
"""Analytics service for topic coverage and gap analysis."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from packages.analytics.coverage import (
    TopicCoverage,
    TopicGap,
    WeakNote,
    get_coverage_tree,
    get_topic_coverage,
    get_topic_gaps,
    get_weak_notes,
)
from packages.analytics.labeling import LabelingStats, label_all_notes
from packages.analytics.taxonomy import (
    Taxonomy,
    load_taxonomy_from_database,
    load_taxonomy_from_yaml,
    sync_taxonomy_to_database,
)
from packages.common.config import Settings, get_settings

if TYPE_CHECKING:
    from pathlib import Path


@dataclass(frozen=True, slots=True)
class AnalyticsResult:
    """Result from analytics operations."""

    coverage: TopicCoverage | None = None
    gaps: list[TopicGap] | None = None
    weak_notes: list[WeakNote] | None = None


class AnalyticsService:
    """Service for topic analytics."""

    def __init__(self, settings: Settings | None = None) -> None:
        self.settings = settings or get_settings()

    async def load_taxonomy(self, yaml_path: Path | None = None) -> Taxonomy:
        """Load taxonomy from YAML or database, syncing to DB if from YAML."""
        if yaml_path:
            taxonomy = load_taxonomy_from_yaml(yaml_path)
            # Sync to database
            await sync_taxonomy_to_database(taxonomy, self.settings)
            # Reload to get IDs
            return await load_taxonomy_from_database(self.settings)
        return await load_taxonomy_from_database(self.settings)

    async def label_notes(
        self,
        taxonomy: Taxonomy | None = None,
        min_confidence: float = 0.3,
    ) -> LabelingStats:
        """Label all notes with topics, loading taxonomy from DB if not provided."""
        if taxonomy is None:
            taxonomy = await load_taxonomy_from_database(self.settings)

        return await label_all_notes(taxonomy, self.settings, min_confidence)

    async def get_coverage(
        self,
        topic_path: str,
        include_subtree: bool = True,
    ) -> TopicCoverage | None:
        return await get_topic_coverage(topic_path, include_subtree, self.settings)

    async def get_gaps(
        self,
        topic_path: str,
        min_coverage: int = 1,
    ) -> list[TopicGap]:
        return await get_topic_gaps(topic_path, min_coverage, self.settings)

    async def get_weak_notes(
        self,
        topic_path: str,
        max_results: int = 20,
    ) -> list[WeakNote]:
        return await get_weak_notes(topic_path, max_results, settings=self.settings)

    async def get_full_analysis(
        self,
        topic_path: str,
        min_coverage: int = 1,
    ) -> AnalyticsResult:
        coverage = await self.get_coverage(topic_path)
        gaps = await self.get_gaps(topic_path, min_coverage)
        weak = await self.get_weak_notes(topic_path)

        return AnalyticsResult(
            coverage=coverage,
            gaps=gaps,
            weak_notes=weak,
        )

    async def get_taxonomy_tree(
        self,
        root_path: str | None = None,
    ) -> list[dict[str, Any]]:
        return await get_coverage_tree(root_path, self.settings)

File: crates/analytics/src/service.rs
use std::sync::Arc;
use tracing::instrument;

use crate::AnalyticsError;
use crate::coverage::{TopicCoverage, TopicGap, WeakNote};
use crate::duplicates::{DuplicateCluster, DuplicateDetector, DuplicateStats};
use crate::labeling::{LabelingStats, TopicLabeler};
use crate::repository::AnalyticsRepository;
use crate::taxonomy::Taxonomy;

/// Facade aggregating taxonomy, coverage, labeling, and duplicate detection.
pub struct AnalyticsService<E, V>
where
    E: indexer::embeddings::EmbeddingProvider,
    V: indexer::qdrant::VectorRepository,
{
    pub embedding: E,
    pub vector_repo: V,
    pub repository: Arc<dyn AnalyticsRepository>,
}

impl<E, V> AnalyticsService<E, V>
where
    E: indexer::embeddings::EmbeddingProvider,
    V: indexer::qdrant::VectorRepository,
{
    pub fn new(embedding: E, vector_repo: V, repository: Arc<dyn AnalyticsRepository>) -> Self {
        Self {
            embedding,
            vector_repo,
            repository,
        }
    }

    /// Load taxonomy from YAML (syncing to DB) or from DB.
    #[instrument(skip(self))]
    pub async fn load_taxonomy(
        &self,
        yaml_path: Option<&std::path::Path>,
    ) -> Result<Taxonomy, AnalyticsError> {
        self.repository.load_taxonomy(yaml_path).await
    }

    /// Label all notes with topics.
    #[instrument(skip(self))]
    pub async fn label_notes(
        &self,
        taxonomy: Option<&Taxonomy>,
        min_confidence: f32,
    ) -> Result<LabelingStats, AnalyticsError> {
        let owned_taxonomy;
        let tax = match taxonomy {
            Some(t) => t,
            None => {
                owned_taxonomy = self.repository.load_taxonomy(None).await?;
                &owned_taxonomy
            }
        };

        TopicLabeler::new(&self.embedding, Arc::clone(&self.repository))
            .label_notes(tax, min_confidence, 5, 100)
            .await
    }

    /// Get coverage metrics for a topic.
    #[instrument(skip(self))]
    pub async fn get_coverage(
        &self,
        topic_path: &str,
        include_subtree: bool,
    ) -> Result<Option<TopicCoverage>, AnalyticsError> {
        self.repository
            .get_topic_coverage(topic_path, include_subtree)
            .await
    }

    /// Find gaps in topic coverage under a root path.
    #[instrument(skip(self))]
    pub async fn get_gaps(
        &self,
        topic_path: &str,
        min_coverage: i64,
    ) -> Result<Vec<TopicGap>, AnalyticsError> {
        self.repository
            .get_topic_gaps(topic_path, min_coverage)
            .await
    }

    /// Get weak notes (high lapse rate) in a topic subtree.
    #[instrument(skip(self))]
    pub async fn get_weak_notes(
        &self,
        topic_path: &str,
        max_results: i64,
    ) -> Result<Vec<WeakNote>, AnalyticsError> {
        self.repository
            .get_weak_notes(topic_path, max_results, 0.1)
            .await
    }

    /// Find clusters of near-duplicate notes.
    #[instrument(skip(self))]
    pub async fn find_duplicates(
        &self,
        threshold: f64,
        max_clusters: usize,
        deck_filter: Option<&[String]>,
        tag_filter: Option<&[String]>,
    ) -> Result<(Vec<DuplicateCluster>, DuplicateStats), AnalyticsError> {
        DuplicateDetector::new(&self.vector_repo, Arc::clone(&self.repository))
            .find_duplicates(threshold, max_clusters, deck_filter, tag_filter)
            .await
    }

    /// Get coverage tree for all topics (optionally filtered by root path).
    #[instrument(skip(self))]
    pub async fn get_taxonomy_tree(
        &self,
        root_path: Option<&str>,
    ) -> Result<Vec<serde_json::Value>, AnalyticsError> {
        self.repository.get_coverage_tree(root_path).await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    common::assert_send_sync!(
        AnalyticsError,
        crate::taxonomy::Topic,
        crate::taxonomy::Taxonomy,
        TopicCoverage,
        TopicGap,
        WeakNote,
        LabelingStats,
        DuplicateCluster,
        DuplicateStats,
    );

    #[test]
    fn analytics_service_is_generic() {
        // Verify the service compiles with mock types
        // (this is a compile-time check, the test body doesn't matter)
        fn _check<
            E: indexer::embeddings::EmbeddingProvider,
            V: indexer::qdrant::VectorRepository,
        >() {
            fn _assert<T: Send + Sync>() {}
            _assert::<AnalyticsService<E, V>>();
        }
    }
}
